collections/_article/rendering-acyclic-graph.md

```diff
@@ -0,0 +1,277 @@
+
+Just as draw lists were naturally mapped to render passes, so too was the fact that the start and end action had to be specified for the textures associated with them. These actions cover operations such as loading the contents of a texture, discarding it or clearing it with a specific color. There's plenty of performance to be gained by choosing the optimal option in each render pass. However, due to some internal design decisions, the introduction of a "CONTINUE" action was made that doesn't actually exist in Vulkan. This action essentially meant "the previous draw list is compatible and the texture should not be transitioned to a different state". For an engine like Godot, which must provide a lot of **optional** post-processing effects and drawing layers, this proved to be a *maintenance nightmare*, ending in a ton of branching in the code that was not very easy to follow and prone to mistakes.
+
+#### Draw list storage textures
+
+Since the device doesn't know if the result of a draw list will be used in a compute pass later, it was necessary to specify which textures needed to transition to the "storage" memory layout at the end of the render pass. *Storage* is the layout required for modifying textures directly from compute passes. From the point of view of an outsider to Godot, there was no clear reason why this had to be part of the API, but when analyzing the implementation it was evident it was introduced to deal with the barrier transitions required by Vulkan. In turn, this delegated the responsibility to the programmer to keep track of whether the textures would be used later in the compute passes in order to handle them efficiently.
+
+#### Draw and compute list overlap
+
+Draw lists and compute lists were allowed to specify whether they can "overlap" during execution. This is a very vague argument that depends on whether the programmer correctly understands whether the render and compute work aren't dependent on each other and could realistically run in parallel. With no real way to validate this, this flag essentially meant "turn it on and hope for the best" as it simply skipped some synchronization steps that the default behavior enforced. Once again, making use of this feature introduced some issues in the codebase where it would require tracking whether parallel work was being submitted or not: components like GI could make use of this optimization, but whether the feature is used or not depends on what that the user has enabled in the scene.
+
+#### Barrier mask
+
+While Rendering Device does not provide explicit access to barriers, a lot of methods would allow specifying a bit mask to define which stages of the pipeline the command must synchronize with. However, this was probably one of the most misused masks in the existing rendering code as its implementation wasn't even consistent in most methods. In one instance, a variable that wasn't a bitmask at all was accidentally casted as one and passed to a rendering method.
```

Comment:
```suggestion
While RenderingDevice does not provide explicit access to barriers, a lot of methods would allow specifying a bit mask to define which stages of the pipeline the command must synchronize with. However, this was probably one of the most misused masks in the existing rendering code as its implementation wasn't even consistent in most methods. In one instance, a variable that wasn't a bitmask at all was accidentally casted as one and passed to a rendering method.
```